collections/iter.go

```diff
@@ -0,0 +1,298 @@
+func BoundInclusive[K any](key K) *Bound[K] {
+	return &Bound[K]{
+		value:     key,
+		inclusive: true,
+	}
+}
+
+// BoundExclusive creates a Bound of the provided key K
+// which is exclusive. Meaning, if it is used as Ranger.RangeValues start,
+// the provided key will be excluded if it exists in the Iterator range.
+func BoundExclusive[K any](key K) *Bound[K] {
+	return &Bound[K]{
+		value:     key,
+		inclusive: false,
+	}
+}
+
+// Bound defines key bounds for Start and Ends of iterator ranges.
+type Bound[K any] struct {
+	value     K
+	inclusive bool
+}
+
+// Ranger defines a generic interface that provides a range of keys.
+type Ranger[K any] interface {
+	// RangeValues is defined by Ranger implementers.
+	// It provides instructions to generate an Iterator instance.
+	// If prefix is not nil, then the Iterator will return only the keys which start
+	// with the given prefix.
+	// If start is not nil, then the Iterator will return only keys which are greater than the provided start
+	// or greater equal depending on the bound is inclusive or exclusive.
+	// If end is not nil, then the Iterator will return only keys which are smaller than the provided end
+	// or smaller equal depending on the bound is inclusive or exclusive.
+	RangeValues() (prefix *K, start *Bound[K], end *Bound[K], order Order, err error)
+}
+
+// Range is a Ranger implementer.
+type Range[K any] struct {
+	prefix *K
+	start  *Bound[K]
+	end    *Bound[K]
+	order  Order
+}
+
+// Prefix sets a fixed prefix for the key range.
+func (r *Range[K]) Prefix(key K) *Range[K] {
+	r.prefix = &key
+	return r
+}
+
+// StartInclusive makes the range contain only keys which are bigger or equal to the provided start K.
+func (r *Range[K]) StartInclusive(start K) *Range[K] {
+	r.start = BoundInclusive(start)
+	return r
+}
+
+// StartExclusive makes the range contain only keys which are bigger to the provided start K.
+func (r *Range[K]) StartExclusive(start K) *Range[K] {
+	r.start = BoundExclusive(start)
+	return r
+}
+
+// EndInclusive makes the range contain only keys which are smaller or equal to the provided end K.
+func (r *Range[K]) EndInclusive(end K) *Range[K] {
+	r.end = BoundInclusive(end)
+	return r
+}
+
+// EndExclusive makes the range contain only keys which are smaller to the provided end K.
+func (r *Range[K]) EndExclusive(end K) *Range[K] {
+	r.end = BoundExclusive(end)
+	return r
+}
+
+func (r *Range[K]) Descending() *Range[K] {
+	r.order = OrderDescending
+	return r
+}
+
+func (r *Range[K]) RangeValues() (prefix *K, start *Bound[K], end *Bound[K], order Order, err error) {
+	return r.prefix, r.start, r.end, r.order, nil
+}
+
+// iteratorFromRanger generates an Iterator instance, with the proper prefixing and ranging.
+// a nil Ranger can be seen as an ascending iteration over all the possible keys.
+func iteratorFromRanger[K, V any](ctx context.Context, m Map[K, V], r Ranger[K]) (iter Iterator[K, V], err error) {
+	// defaults
+	var (
+		prefix *K
```

Comment:
don't like `prefix` naming as it is confusing with `Prefix`.
the latter defines a storage object bytes prefix 
the former defines an iteration over a key prefix

Reply:
I understand, it's jsut that when I was reviewing my own code I read this field: https://github.com/cosmos/cosmos-sdk/pull/14222/files#diff-3eaa8234d537015aca3c6f9aee92453897778031b44159500b0005462190b8dbR202

and i was like, am I using the Range.Prefix or the `Prefix`? I had to comment it in order to clarify it, which lead me to think if we should  have 2 different names.

Reply:
We can just call the other one `collectionPrefix` for now maybe